### PR TITLE
new module atlas_ldap_user for DB Users authenticated via LDAP

### PR DESCRIPTION
**Pull request by derekgiri**
… AD Groups and the password parameter is not enforced but an option for SCRAM password based users

### CODE REVIEW

**Comment by derekgiri** [comment]
Rationale:
1. Users typically form part of an AD group and so when creating users in MongoDB, you essentially create a distinguished name where the common name matches the AD group name. The DN is also stored in the admin database within MongoDB.
2. Create users with SCRAM based authentication is also possible, the ldapAuthType can be excluded in the task and instead replaced by the password parameter. 

Both the ldapAuthType and password parameters can be used one instead of another but not both in the ansible task. This is because users with SCRAM based auth cannot use LDAP auth as well.

**Comment by derekgiri** [comment]
Multiple Auth types cannot be specified as shown below as this will error when running the ansible task. 
Hence it necessary to create a separate module for creating users authenticated via LDAP.

   
- name: test user
      atlas_user:
        apiUsername: "API_user"
        apiPassword: "API_passwort_or_token"
        groupId: "GROUP_ID"
        ldapAuthType: "GROUP"
        username: my_app_user
        password: SuperSecret!
        roles:
          - databaseName: private_info
            roleName: read
          - databaseName: public_info
            roleName: readWrite

File: plugins/modules/atlas_ldap_user.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+DOCUMENTATION = """
+---
+module: atlas_ldap_user
+short_description: Manage database users in Atlas
+description:
+   - The atlas_users module lets you create, modify and delete the database users in your cluster.
```
Comment (schurzi) [inline plugins/modules/atlas_ldap_user.py:35]:
```suggestion
   - The atlas_ldap_user module lets you create, modify and delete the database users in your cluster.
```

File: plugins/modules/atlas_ldap_user.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+   - Each user has a set of roles that provide access to the project's databases.
+   - A user's roles apply to all the clusters in the project
+   - if two clusters have a products database and a user has a role granting read access on the products database,
+   - the user has that access on both clusters.
+   - L(API Documentation,https://docs.atlas.mongodb.com/reference/api/database-users/)
+author: "Martin Schurz (@schurzi)"
```
Comment (schurzi) [inline plugins/modules/atlas_ldap_user.py:41]: please add your author info here :)

File: plugins/modules/atlas_ldap_user.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      - State of the ressource.
+    choices: [ "present", "absent" ]
+    default: present
+    type: str
+  groupId:
+    description:
+      - Unique identifier for the Atlas project.
+    type: str
+    required: True
+  databaseName:
```
Comment (schurzi) [inline plugins/modules/atlas_ldap_user.py:66]: this should be hardcoded to external, or am I understanding the documentation wrong?

**Review by schurzi — CHANGES_REQUESTED** [review]
Thank you for his great addition. I think the current approach with an extra module is good, since it does not create possibilities for problems with different kinds of users.

Could you please fix the three little issues and possibly add some extra text to the desctiption about usage of ldap for authentication?

File: plugins/modules/atlas_ldap_user.py (hunk 3 above)
Comment (derekgiri) [reply plugins/modules/atlas_ldap_user.py:66]: If you enable user authorization with LDAP, you can create LDAP groups on the admin database by mapping LDAP groups to MongoDB roles on your Atlas databases. To use LDAP groups effectively, create additional projects within Atlas to control access to specific deployments in your organization, such as creating separate Atlas projects for development and production environments. You can then map an LDAP group to a role in the Atlas project to provide access to the desired deployment.

Comment (schurzi) [reply plugins/modules/atlas_ldap_user.py:63]: ahh ok, so I misunderstood. Only the connecting client needs to specify `$external`

**Review by schurzi — APPROVED** [review]

**Pull request merged**